engines/e_cng.c

```diff
@@ -0,0 +1,1586 @@
+                ctx->errcnt++;
+                return NULL;
+            }
+        }
+
+        STORE_INFO_free(ret);
+        if ((ret = meth->load(ctx->data)) == NULL) {
+            if (meth->error(ctx->data))
+                ctx->errcnt++;
+            if (!meth->clean(ctx->data)) {
+                STORE_INFO_free(ret);
+                ctx->errcnt++;
+            }
+            if (cng_store_error(ctx))
+                return NULL;
+
+            /* OK, we reached EOF for this method, time to try the next */
+            ctx->data = NULL;
+            (void)sk_STORE_LOOKUP_FNS_shift(ctx->meths);
+            goto again;
+        }
+
+    }
+    return ret;
+}
+
+int cng_store_eof(STORE_LOADER_CTX *ctx)
+{
+    return sk_STORE_LOOKUP_FNS_num(ctx->meths) == 0;
+}
+
+int cng_store_error(STORE_LOADER_CTX *ctx)
+{
+    return ctx->errcnt > 0;
+}
+
+int cng_store_close(STORE_LOADER_CTX *ctx)
+{
+    OPENSSL_free(ctx);
+    return 1;
+}
+
+static STORE_LOADER *cng_store_new(ENGINE *e)
+{
+    STORE_LOADER *res = NULL;
+
+    if ((res = STORE_LOADER_new(e, cng_scheme)) == NULL
+        || !STORE_LOADER_set_open(res, cng_store_open)
+        || !STORE_LOADER_set_load(res, cng_store_load)
+        || !STORE_LOADER_set_eof(res, cng_store_eof)
+        || !STORE_LOADER_set_error(res, cng_store_error)
+        || !STORE_LOADER_set_close(res, cng_store_close)) {
+        CNGerr(CNG_F_CNG_STORE_NEW, ERR_R_MALLOC_FAILURE);
+        STORE_LOADER_free(res);
+        res = NULL;
+    }
+    return res;
+}
+
+static void cng_store_free(STORE_LOADER *loader)
+{
+    STORE_LOADER_free(loader);
+}
+
+/*-
+ * RSA key method
+ * --------------
+ */
+
+static CRYPTO_REF_COUNT rsa_cnt = 0;
+static CRYPTO_RWLOCK *rsa_cnt_lock = NULL;
+static CRYPTO_ONCE rsa_cnt_init = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_rsa_cnt_init)
+{
+    OPENSSL_init_crypto(0, NULL);
+    rsa_cnt_lock = CRYPTO_THREAD_lock_new();
+    return rsa_cnt_lock != NULL;
+}
+
+static int ossl_to_cng_rsa_padding(int ossl_padding)
+{
+    switch (ossl_padding) {
+    case RSA_PKCS1_PADDING:
+        return NCRYPT_PAD_PKCS1_FLAG;
+    case RSA_NO_PADDING:
+        return NCRYPT_NO_PADDING_FLAG;
+# if 0                           /* Currently unsupported */
+    case RSA_PKCS1_OAEP_PADDING:
+        return NCRYPT_PAD_OAEP_FLAG;
+# endif
+    default:
+        break;
+    }
+    return -1;
+}
+
+static int cng_rsa_priv_enc(int flen, const unsigned char *from,
+                            unsigned char *to, RSA *rsa, int padding)
+{
+    CNG_CTX *ctx = RSA_get_ex_data(rsa, rsa_idx);
+    int pad = ossl_to_cng_rsa_padding(padding);
+    DWORD len;
+
+    if (pad == -1) {
+        char errstr[10];
+        BIO_snprintf(errstr, 10, "%d", padding);
+        CNGerr(CNG_F_CNG_RSA_PRIV_ENC, CNG_R_UNSUPPORTED_PADDING);
+        ERR_add_error_data(2, "padding=", errstr);
+        return -1;
+    }
+
+    if (NCryptEncrypt(ctx->khandle, (PBYTE)from, flen, NULL, to, flen, &len,
```

Comment:
As far as I have seen, `NCryptEncrypt` will use the public key from the key handle instead of the private key, so `NCryptDecrypt` needs to be used here to perform the RSA operation with the private key.

`NCryptDecrypt` however probably can't be configured to add padding before the RSA operation with the private key as required here (RSA private enrypt) as it's designed to be used for decryption with the private key but not for signing/encryption and therefore removes the padding after the RSA operation (RSA private decrypt) instead of adding it before.
For that reason, `NCryptDecrypt` can only be used with `RSA_NO_PADDING` here, as with that RSA private encrypt and RSA private decrypt result in the same operation.

When OpenSSL uses RSA signing with PSS padding, this method is called with `RSA_NO_PADDING` as the padding has already be applied to the buffer before, so that is no problem in that case.

Client certificate authentication with TLS 1.2+ works in my test case, when `NCryptDecrypt` with `RSA_NO_PADDING` is used here (See https://github.com/openssl/openssl/issues/12859).

Please tell me if I am wrong.

Reply:
Yes, it is true that from my testing that when using TLSv1.3, the `NCryptDecrypt` should be used instead of `NCryptEncrypt` with the `NCRYPT_NO_PADDING_FLAG` as the padding of the buffer.